SQL: Failing Group By queries due to different ExpressionIds

emasab opened the pull request:
Fix an issue that arises from the use of ExpressionIds as keys in a lookup map
that helps the `QueryTranslator` to identify the grouping columns. The issue is
that the same expression in different parts of the query (SELECT clause and GROUP BY clause) 
ends up with different ExpressionIds so the lookup fails. So, instead of ExpressionIds
use the hashCode() of NamedExpression.

Fixes: #41159
Fixes: #40001
Fixes: #40240
Fixes: #33361
Fixes: #46316
Fixes: #36074
Fixes: #34543
Fixes: #37044

Fixes: #42041

emasab committed:
bugfix for https://github.com/elastic/elasticsearch/issues/42041

elasticmachine commented:
Pinging @elastic/es-search

matriv reviewed (commented):
Hello @emasab!

First of all apologies for the delay on picking up this PR and many thanks for your effort here.
The issue with the ExpressionIds is quite complicated and there are plans to solve it overall through a substantial refactoring. Your changes though, solve a substantial subset of the issues 
so we decided to go ahead and proceed with this fix. I've changed the title and description of the PR and link to the issues that it fixes.

For some cases (without aliases or numeric refs) queries still fail, e.g.:
```SELECT PI() * emp_no FROM test_emp GROUP BY PI() * emp_no ORDER BY PI() * emp_no LIMIT 10```

Moreover I'd like to ask from you to add some integration tests for the issues that this fix addresses:
In `agg.sql-spec` file: https://github.com/emasab/elasticsearch/blob/bugfix/42041-failing-group-by-queries/x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec#L52 the following:
```
groupByCastScalar
SELECT CAST(ABS(EXTRACT(YEAR FROM "birth_date")) AS BIGINT) FROM test_emp GROUP BY CAST(ABS(EXTRACT(YEAR FROM "birth_date")) AS BIGINT) ORDER BY CAST(ABS(EXTRACT(YEAR FROM "birth_date")) AS BIGINT) NULLS FIRST;
groupByCastScalarWithAlias
SELECT CAST(ABS(EXTRACT(YEAR FROM "birth_date")) AS BIGINT) as "cast" FROM test_emp GROUP BY "cast" ORDER BY "cast" NULLS FIRST;
```
and in `agg.csv-spec` file: https://github.com/emasab/elasticsearch/blob/bugfix/42041-failing-group-by-queries/x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec#L220
the following:
```
// Fails for H2
groupByCastScalarWithNumericRef
SELECT CAST(ABS(EXTRACT(YEAR FROM "birth_date")) AS BIGINT) FROM test_emp GROUP BY 1 ORDER BY 1 NULLS FIRST;

CAST(ABS(EXTRACT(YEAR FROM "birth_date")) AS BIGINT):l
------------------------------------------------------
null
1952
1953
1954
1955
1956
1957
1958
1959
1960
1961
1962
1963
1964
1965
;

groupByConvertScalar
SELECT CONVERT(ABS(EXTRACT(YEAR FROM "birth_date")), SQL_BIGINT) FROM test_emp GROUP BY CONVERT(ABS(EXTRACT(YEAR FROM "birth_date")), SQL_BIGINT) ORDER BY CONVERT(ABS(EXTRACT(YEAR FROM "birth_date")), SQL_BIGINT) NULLS FIRST;


CONVERT(ABS(EXTRACT(YEAR FROM "birth_date")), SQL_BIGINT):l
-----------------------------------------------------------
null
1952
1953
1954
1955
1956
1957
1958
1959
1960
1961
1962
1963
1964
1965
;


groupByConvertScalarWithAlias
SELECT CONVERT(ABS(EXTRACT(YEAR FROM "birth_date")), SQL_BIGINT) as "convert" FROM test_emp GROUP BY "convert" ORDER BY "convert" NULLS FIRST;

convert:l
---------
null
1952
1953
1954
1955
1956
1957
1958
1959
1960
1961
1962
1963
1964
1965
;

groupByConvertScalarWithNumericRef
SELECT CONVERT(ABS(EXTRACT(YEAR FROM "birth_date")), SQL_BIGINT) FROM test_emp GROUP BY 1 ORDER BY 1 NULLS FIRST;

CONVERT(ABS(EXTRACT(YEAR FROM "birth_date")), SQL_BIGINT):l
-----------------------------------------------------------
null
1952
1953
1954
1955
1956
1957
1958
1959
1960
1961
1962
1963
1964
1965
;

// AwaitsFix https://github.com/elastic/elasticsearch/issues/36074
//groupByConstantScalar
//SELECT PI() * emp_no FROM test_emp GROUP BY PI() * emp_no ORDER BY PI() * emp_no LIMIT 10;

//PI() * emp_no:d
//---------------
//31419.0681285515
//31422.2097212051
//31425.3513138587
//31428.4929065123
//31431.6344991659
//31434.7760918195
//31437.9176844731
//31441.0592771266
//31444.2008697802
//31447.3424624338
//;

groupByConstantScalarWithAlias
SELECT PI() * emp_no AS "value" FROM test_emp GROUP BY value ORDER BY value LIMIT 10;

value:d
-------
31419.0681285515
31422.2097212051
31425.3513138587
31428.4929065123
31431.6344991659
31434.7760918195
31437.9176844731
31441.0592771266
31444.2008697802
31447.3424624338
;

groupByConstantScalarWithNumericRef
SELECT PI() * emp_no FROM test_emp GROUP BY 1 ORDER BY 1 LIMIT 10;

PI() * emp_no:d
---------------
31419.0681285515
31422.2097212051
31425.3513138587
31428.4929065123
31431.6344991659
31434.7760918195
31437.9176844731
31441.0592771266
31444.2008697802
31447.3424624338
;
```

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:
```diff
@@ -779,6 +779,85 @@ public void testAllCountVariantsWithHavingGenerateCorrectAggregations() {
                         + "\"gap_policy\":\"skip\"}}}}}"));
     }
 
+    public void testGroupSelection() {
```
It would be great to split this up to more methods with more descriptive names and add more tests so that all 3 cases (no alias, alias, numeric reference) are covered, similarly to the integration tests in my comment: https://github.com/elastic/elasticsearch/pull/43072#pullrequestreview-267772000
Of course some of the cases, as mentioned, should be skipped as they won't work even with this fix.

emasab commented:
Hi @matriv , happy to hear from you. I understand, the checks to do are a lot before giving an ok. I'm going to include these integration tests and see if there is a simple fix for the failing query that you've highlighted.

emasab committed:
added integration tests
added test cases and refactored into multiple unit tests

emasab committed:
fix for for query with "group by" and "order by" a constant scalar
unified semantic of equals to that of hashCode for NamedExpression
changed AttributeMapTests according to the new semantic
changed equals semantic of Failure

emasab commented:
I've fixed the groupByConstantScalar case too, mainly unifying the semantic of NamedExpression.equals to that of NamedExpression.hashCode . The ExpressionId is still there but is used only as a unique identifier for the object instance. I've changed the semantic of Failure.equals because now two NamedExpression are equal if they have the same tree even if the position in the source is different so Failure takes into account the source position not the parsed node.

emasab commented:
Sorry, the latest commit

https://github.com/elastic/elasticsearch/pull/43072/commits/490201553ecff596ea5613dd3a4d616ed84ba094

even if more correct theorically because of the difference between hashCode and equals, solves the specified case and passes the unit tests, causes a lot of failures with the integration test org.elasticsearch.xpack.sql.qa.single_node.JdbcSqlSpecIT. The previous one 

https://github.com/elastic/elasticsearch/pull/43072/commits/87205638b7dd46e0ac7a9894ebee4a99561fac5b

is not affected

emasab committed:
Changed the behavior of QueryFolder and QueryContainer to use ExpressionId.
Minor changes to adapt to the change of NamedExpression.equals and hashCode.
Changed unit test because of the same reason.

emasab commented:
In the latest commit I've solved the integration tests failures while keeping the equals and hashCode methods consistent. The order by constant scalar is still excluded, With my previous commit there was an incomplete change to ResolveMissingRefs that made it work but I've reverted it because this other query failed:

`SELECT last_name FROM "test_emp" ORDER BY birth_date::TIME, emp_no LIMIT 5`

emasab committed:
fix for order by and group by ScalarFunctionAttribute

emasab commented:
In the latest commit these is a fix for groupByConstantScalar case with ORDER BY, while keeping the equals consistent with the hashCode, and the integration tests successful

emasab committed:
changed excepted test behavior to be similiar to that of testGeoShapeInWhereClause

emasab commented:
rebased on current master

emasab committed:
fix order by grouped ScalarFunctionAttribute
rebased

emasab commented:
In the latest commit there is a fix for the results ordering when there is a grouped ScalarFunctionAttribute in the ORDER BY clause, like:
`
SELECT PI() * emp_no FROM test_emp GROUP BY PI() * emp_no ORDER BY PI() * emp_no DESC LIMIT 10;`

emasab committed:
removed unused import

emasab commented:
rebased

emasab committed:
solved recent issue with Pivot and EsRelation

emasab commented:
solved issue with Pivot query and rebased

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java:
```diff
@@ -102,6 +103,32 @@ public static AttributeSet references(List<? extends Expression> exps) {
     }
 
+    public static AttributeSet referencesExcept(Set<? extends Expression> exps,
+                                                AttributeSet except) {
```
could be on the same line I guess?

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/gen/script/Param.java:
```diff
@@ -24,4 +26,18 @@ T value() {
+
+    @Override
+    public int hashCode() {
+        if(this.value==null) return Objects.hashCode(null);
```
```suggestion
        if (this.value == null) {
            return Objects.hashCode(null);
        }
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/gen/script/Param.java:
```diff
@@ -24,4 +26,18 @@ T value() {
+
+    @Override
+    public boolean equals(Object obj) {
+        if(!(obj instanceof  Param)) return false;
```
```suggestion
        if ((obj instanceof  Param) == false) {
            return false;
        }
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/gen/script/Param.java:
```diff
@@ -24,4 +26,18 @@ T value() {
+    @Override
+    public boolean equals(Object obj) {
+        if(!(obj instanceof  Param)) return false;
+        if(this.value==null) return ((Param)obj).value == null;
```
```suggestion
        if (this.value == null) {
            return ((Param)obj).value == null;
        }
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java:
```diff
@@ -109,7 +110,7 @@ public QueryContainer(Query query,
         this.aggs = aggs == null ? Aggs.EMPTY : aggs;
         this.fields = fields == null || fields.isEmpty() ? emptyList() : fields;
-        this.aliases = aliases == null || aliases.isEmpty() ? AttributeMap.emptyAttributeMap() : aliases;
+        this.aliases = aliases == null || aliases.isEmpty() ? new HashMap<>() : aliases;
```
Why not `Collections.emptyMap()`?

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/Aggs.java:
```diff
@@ -121,16 +123,23 @@ public Aggs addAgg(PipelineAgg pipelineAgg) {
+            if (id.equals(group.id())) {
                 return group;
             }
+            if(attr instanceof ScalarFunctionAttribute) {
```
```suggestion
            if (attr instanceof ScalarFunctionAttribute) {
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/Aggs.java:
```diff
@@ -121,16 +123,23 @@ public Aggs addAgg(PipelineAgg pipelineAgg) {
             }
+            if(attr instanceof ScalarFunctionAttribute) {
+                ScalarFunctionAttribute sfa = (ScalarFunctionAttribute) attr;
+                if (group.script()!=null && group.script().equals(sfa.script())) {
```
```suggestion
                if (group.script() !=null && group.script().equals(sfa.script())) {
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/gen/script/Params.java:
```diff
@@ -124,4 +124,15 @@ else if (p instanceof Var) {
+
+    @Override
+    public boolean equals(Object obj) {
+        if(!(obj instanceof  Params)) return false;
```
```suggestion
        if ((obj instanceof  Params) == false) {
            return false;
        }
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java:
```diff
@@ -47,7 +48,20 @@ public Pivot(Source source, LogicalPlan child, Expression column, List<NamedExpr
     protected Pivot replaceChild(LogicalPlan newChild) {
-        return new Pivot(source(), newChild, column, values, aggregates);
+
+       java.util.function.Function<Expression, Expression> withQualifierNull = (Expression e)-> {
```
I would extract this to a private static attribute.

emasab replied:
done

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java:
```diff
@@ -47,7 +48,20 @@ public Pivot(Source source, LogicalPlan child, Expression column, List<NamedExpr
     @Override
     protected Pivot replaceChild(LogicalPlan newChild) {
-        return new Pivot(source(), newChild, column, values, aggregates);
+
```
Could you please add some comment on why this is needed?

emasab replied:
done

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java:
```diff
@@ -102,6 +103,32 @@ public static AttributeSet references(List<? extends Expression> exps) {
+    public static AttributeSet referencesExcept(Set<? extends Expression> exps,
+                                                AttributeSet except) {
+        AttributeSet ret = new AttributeSet();
+        while(exps.size()>0){
```
I don't get the logic here, why do you need the while loop and revisit the `exps` which have become `filteredExps`?

emasab replied:
It's a breadth-first search where some nodes are removed during the traversal. At each loop the exps represents the previous list of tree nodes and filteredExps the expressions that aren't in the except set. At the end of the loop exps is replaced with the children of each node. When a leaf is found, its references are added to ret.

matriv commented:
@elasticmachine test this please

matriv reviewed (commented):
Hello @emasab,

Once again thanks a lot for your effort here.
There has been delay in getting back to you, as there were other features and bugs in our plate, and as stated before there is an ongoing effort of complete refactoring in the area of NamedExpression and ExpressionIds.

The state of the PR though looks good overall to me and would be nice to fix those issues independently of the refactoring that might take more time to be ready.

I left another round of comments and I'd like to do some more testing myself using your branch.

emasab committed:
Apply suggestions from code review

code style

Co-Authored-By: Marios Trivyzas <matriv@gmail.com>

emasab committed:
changes suggested by @matriv

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java:
```diff
@@ -40,14 +41,38 @@ public Pivot(Source source, LogicalPlan child, Expression column, List<NamedExpr
+        Expression newColumn = column;
+        List<NamedExpression> newAggregates = aggregates;
+
+        if(newChild instanceof EsRelation) {
```
```suggestion
        if (newChild instanceof EsRelation) {
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java:
```diff
@@ -40,14 +41,38 @@ public Pivot(Source source, LogicalPlan child, Expression column, List<NamedExpr
         this.aggregates = aggregates;
     }
 
+    private static Expression withQualifierNull(Expression e){
```
```suggestion
    private static Expression withQualifierNull(Expression e) {
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/Pivot.java:
```diff
@@ -40,14 +41,38 @@ public Pivot(Source source, LogicalPlan child, Expression column, List<NamedExpr
     @Override
     protected Pivot replaceChild(LogicalPlan newChild) {
-        return new Pivot(source(), newChild, column, values, aggregates);
+
```
```suggestion
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java:
```diff
@@ -102,6 +103,31 @@ public static AttributeSet references(List<? extends Expression> exps) {
+        while(exps.size()>0){
+
+            Set<Expression> filteredExps = new LinkedHashSet<>();
+            for(Expression exp: exps){
```
```suggestion
            for (Expression exp : exps) {
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java:
```diff
@@ -102,6 +103,31 @@ public static AttributeSet references(List<? extends Expression> exps) {
+            Set<Expression> filteredExps = new LinkedHashSet<>();
+            for(Expression exp: exps){
+                Expression attr = Expressions.attribute(exp);
+                if(attr==null || !except.contains(attr)){
```
```suggestion
                if (attr == null || (except.contains(attr) == false)) {
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java:
```diff
@@ -102,6 +103,31 @@ public static AttributeSet references(List<? extends Expression> exps) {
 
+    public static AttributeSet referencesExcept(Set<? extends Expression> exps, AttributeSet except) {
+        AttributeSet ret = new AttributeSet();
+        while(exps.size()>0){
```
```suggestion
        while (exps.size() > 0) {
```

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expressions.java:
```diff
@@ -102,6 +103,31 @@ public static AttributeSet references(List<? extends Expression> exps) {
         return set;
     }
 
+    public static AttributeSet referencesExcept(Set<? extends Expression> exps, AttributeSet except) {
```
Maybe a name with a verb, something `filterReferences` ?

matriv replied:
I would name it `excluded`.

emasab replied:
ok

matriv reviewed (commented):
Thx for addressing the comments @emasab 
I left a few more, minor, regarding code formatting.

matriv commented on x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec:
```diff
@@ -246,6 +246,194 @@ TRUNCATE(YEAR("birth_date"), -2)
+M           |4.1415926535
+;
+
+groupByFieldAndConstantScalarWithAliasWithOrderByDescWithNullsFirst
```
@emasab please don't forget to check this.

emasab committed:
Apply suggestions from code review

code style

Co-Authored-By: Marios Trivyzas <matriv@gmail.com>

emasab committed:
change test case name

emasab commented:
> Thx for addressing the comments @emasab
> I left a few more, minor, regarding code formatting.

I've set up the save action and the eclipseformat :)

emasab committed:
referencesExcept changed to filterReferences

matriv reviewed (approved):
LGTM. Thank you @emasab for your work, your patience and the effort and time spent on addressing the comments.

matriv commented:
@elasticmachine test this please
